python/ql/src/experimental/Security/CWE-287-ConstantSecretKey/WebAppConstantSecretKey.ql

```diff
@@ -0,0 +1,81 @@
+/**
+ * @name Initializing SECRET_KEY of Flask application with Constant value
+ * @description Initializing SECRET_KEY of Flask application with Constant value
+ * files can lead to Authentication bypass
+ * @kind path-problem
+ * @id py/flask-constant-secret-key
+ * @problem.severity error
+ * @security-severity 8.5
+ * @precision high
+ * @tags security
+ *       experimental
+ *       external/cwe/cwe-287
+ */
+
+import python
+import experimental.semmle.python.Concepts
+import semmle.python.dataflow.new.DataFlow
+import semmle.python.ApiGraphs
+import semmle.python.dataflow.new.TaintTracking
+import WebAppConstantSecretKeyDjango
+import WebAppConstantSecretKeyFlask
+
+private predicate stringConstCompare(DataFlow::GuardNode g, ControlFlowNode node, boolean branch) {
+  exists(CompareNode cn | cn = g |
+    exists(StrConst str_const, Cmpop op |
+      op = any(Eq eq) and branch = false
+      or
+      op = any(NotEq ne) and branch = true
+    |
+      cn.operands(str_const.getAFlowNode(), op, node)
+      or
+      cn.operands(node, op, str_const.getAFlowNode())
+    )
+  )
+}
+
+class StringConstCompareBarrier extends DataFlow::Node {
+  StringConstCompareBarrier() {
+    this = DataFlow::BarrierGuard<stringConstCompare/3>::getABarrierNode()
+  }
+}
+
+newtype TFrameWork =
+  Flask() or
+  Django()
+
+module WebAppConstantSecretKeyConfig implements DataFlow::StateConfigSig {
+  class FlowState = TFrameWork;
+
+  predicate isSource(DataFlow::Node source, FlowState state) {
+    state = Flask() and FlaskConstantSecretKeyConfig::isSource(source)
+    or
+    state = Django() and DjangoConstantSecretKeyConfig::isSource(source)
+  }
+
+  predicate isSink(DataFlow::Node sink, FlowState state) {
+    state = Flask() and FlaskConstantSecretKeyConfig::isSink(sink)
+    or
+    state = Django() and DjangoConstantSecretKeyConfig::isSink(sink)
+  }
+
+  predicate isBarrier(DataFlow::Node sanitizer, FlowState state) {
+    (state = Flask() or state = Django()) and
+    sanitizer instanceof StringConstCompareBarrier
+  }
```

Comment:
You wrote the barrier exactly as I described. But it will not work, because we are only tracking flow inside the config files and the source has reached the sink before it gets to this barrier. The barrier, when written as a raise at the bottom of the config file, effectively checks that flow does not leave the config file. This ensures that the value is not read when the config file is loaded, but the query does not see that as there is a complete path from source to sink inside the config file.

I am sorry to have lead you down this path before reading your code properly.

I see three ways to proceed:
1. Use a safe config file that does not need a barrier (one that reads from the environment, perhaps) for the safe example
2. Filter out sinks that are not live at the end of the config file
3. Make the value flow across files, that is, from the end of the file to the load of that file, via an extra flow step